### DISCUSSION TITLE
Automated Testing Ideas

### DISCUSSION THREAD

**Discussion by ZzRanger** [acmutd/comet_planning#90] [Ideas]
As Planner moves from Alpha to Beta release, it is important to ensure that all parts of the application work as intended every time a change is made in the codebase. So far, we have solely relied on manual testing, which is quite time-consuming to perform and is not always reliable, especially as the functionality becomes more and more complex. Thus, the Planner Maintainers want to set up automated tests to save developer hours and improve maintainability for current and future maintainers. 

The purpose of this discussion is to figure out how we want to go about implementing automated tests. Some things to consider would be the type of tests we to include (i.e. unit tests, functional tests), the scope our tests will cover, and frameworks we want to leverage. Feel free to drop any ideas or suggestions in the comments below.

**Comment by rajmeet2001** [comment]
What are your current ideas on this issue?

**Comment by ZzRanger** [comment]
I was thinking about using Cypress.js as an all-in-one framework for running tests. Cypress is easy to set up & use, runs in the browser, allows us to write a variety of tests (unit, integration, E2E), and has excellent support with React & NextJS. I considered Selenium since it's more widely used, more flexible, and has greater cross-browser support (Cypress only supports Chromium + Firefox), but it requires a lot more setup, has a steeper learning curve, and isn't tailored for JavaScript-based applications like Cypress is.

I'd love to hear other people's thoughts on this.

**Reply by rajmeet2001** [reply]
I like the idea behind cypress.js. However, I believe that before deciding on a framework for unit testing, we need to hash out what we need to test. Having a plan would help us identify our specific needs and better help us decide on a framework that meets those needs.

**Reply by ZzRanger** [reply]
Here is a list of things I think we should test off the top of my head. The ones I marked with an asterisk are the ones I think are more pressing

Unit tests:
 - degree plan tool state*
 - plan generator*
 - onboarding
 - general redux tests*

Integration tests:
 - authentication*
 - protected routes
 - API tests

E2E tests:
 - logging in and updating a degree plan

I know we most likely will not be able to set a lot of this up before the March 28th deadline, but I think a lot of this would be helpful to have for maintainers and especially contributors to know that the changes they make won't break anything. Let me know what you think.

**Comment by ZzRanger** [comment]
This issue is relevant to this discussion. 

https://github.com/UTDNebula/planner/issues/223
